### PR TITLE
fix: sshnp exit if atSign DNE, sshnpd wait for atSign to exist

### LINKED ISSUE #188: Sshnp not receiving Sshnpd deviceName until both atSigns activated
### Describe the bug

The sshnpd tmux session continuously runs `sshnpd` and patiently waits for the `.atKeys` to be generated. If the device atSign is onboarded before the client atSign, then the deviceName is never shared with the client atSign, since the client atSign does not yet have a `public:publickey@`

If I am correct, the notification of the deviceName is being sent without caring if the client receives it or not. Then goes straight into `monitor`. Only once the client is setup, then the device's sshnpd has to be restarted

### Steps to reproduce

1. On device, run sshnpd installer
2. On device, run `./at_activate`
3. On device, tmux session successfully started sshnpd, :)
4. On client, run sshnp installer
5. On client, run `./at_activate`
6. On client: Could not receive the device name from sshnpd since sshnpd tried sharing the device name before sshnp's atsign could activate first

### Expected behavior

I guess it is working as intended (in that two non-onboarded atSigns are unable to communicate to one another), but if the goal of the do-while inside the tmux session was for it to start up the daemon, then it is not going as planned :(

`sshnp` recognize deviceName given by `sshnpd`

### Screenshots

_No response_

### Smartphones

_No response_

### Were you using an atApplication when the bug was found?

_No response_

### Additional context

2 fixes:

1. (documentation fix): expect the person to *know* that both of the atSigns MUST be onboarded before running the `sshnpd` and `sshnp` binaries. maybe even avoid starting up the tmux session to let *them* start it themselves? tell them to specifically that client `sshnp` HAVE to be setup before device `sshnpd`

3. (software fix): ensure that the notifications are received and that the receiving-end is indeed an onboarded atSign, by doing `final NotificationResult notifResult = await notificationService.notify(...)` check for `    notificationResult.notificationStatusEnum == NotificationStatusEnum.delivered`

### PR DESCRIPTION
**Pull request by JeremyTubongbanua**
## What I did

1. util function `atSignIsActivated`
2. sshnpd sleeps 5 seconds, if `sshnp` atSign does not exist
3. sshnp exit(1),  if `sshnpd` atSign does not exist

Fixes issue #118 
 
closes #188

### CODE REVIEW

**Comment by JeremyTubongbanua** [comment]
@cconstab 
For sshnpd, if the sshnp atSign does not exist, do we want to exit(1), or do a while() {sleep 5} or something?

File: lib/atsign_exists.dart [hunk 1]
```diff
@@ -0,0 +1,22 @@
+import 'package:at_client/at_client.dart';
+
+/// Checks if the provided atSign's atServer has been properly activated with a public RSA key.
+/// `atClient` must be authenticated
+/// `atSign` is the atSign to check
+Future<bool> atSignIsActivated(final AtClient atClient, String atSign) async {
```
Comment (gkc) [inline lib/atsign_exists.dart]: Can you add into the sshnp_utils.dart please? Keeps things tidy

File: lib/atsign_exists.dart [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+  final AtKey publicKey = AtKey()
+    ..sharedBy = atSign
+    ..key = 'publickey'
+    ..metadata = metadata;
+
+  try {
+    await atClient.get(publicKey);
+    return true;
+  } catch (e) {
```
Comment (gkc) [inline lib/atsign_exists.dart]: Shouldn't you distinguish between errors? For example if the atServer is not reachable for some reason, that is different than getting a KeyNotFound exception

File: lib/sshnp.dart [hunk 3]
```diff
@@ -177,6 +178,10 @@ class SSHNP {
       throw StateError('Cannot init() - already initialized');
     }
 
+    if(!(await atSignIsActivated(atClient, sshnpdAtSign))) {
+        throw ('sshnpd atSign $sshnpdAtSign does not exist');
```
Comment (gkc) [inline lib/sshnp.dart]: With your current implementation, it might exist, but not yet be onboarded

**Review by gkc — CHANGES_REQUESTED** [review]

File: lib/sshnp.dart (hunk 3 above)
Comment (JeremyTubongbanua) [reply lib/sshnp.dart]: The exception message was incorrect
https://github.com/atsign-foundation/sshnoports/pull/211/commits/e2fd4426a83396de5536c59ada6a55d4c3f552b4

File: lib/atsign_exists.dart (hunk 2 above)
Comment (JeremyTubongbanua) [reply lib/atsign_exists.dart]: https://github.com/atsign-foundation/sshnoports/pull/211/commits/51781b9e6ac9b2097660647043776aa23ea38789

Ah, is this what you mean?

The `AtClient.get` exceptions are:

```dart
Throws [AtKeyException] for the invalid key formed
Throws [AtDecryptionException] if fails to decrypt the value
Throws [AtPrivateKeyNotFoundException] if the encryption private key is not found to decrypt the value
Throws [AtPublicKeyChangeException] if the encryption public key used encrypt the value
is different from the current encryption public key(at the time of decryption)
Throws [SharedKeyNotFoundException] if the shared key to decrypt the value is not found
Throws [SelfKeyNotFoundException] if the self encryption key is not found.
Throws [AtClientException] if the cloud secondary is invalid or not reachable
```

The only 2 possible exceptions should be AtKeyException and AtClientException, right?

File: lib/atsign_exists.dart (hunk 1 above)
Comment (JeremyTubongbanua) [reply lib/atsign_exists.dart]: https://github.com/atsign-foundation/sshnoports/pull/211/commits/51781b9e6ac9b2097660647043776aa23ea38789

File: lib/sshnp.dart (hunk 3 above)
Comment (JeremyTubongbanua) [reply lib/sshnp.dart]: Oh, an atSign can have a `publickey` but not be onboarded?

**EDIT**

I find that it does not exist, as long as we bypass cache

```
@jeremy_0 plookup:bypassCache:true:publickey@slow674
=> error:{"errorCode":"AT0011","errorDescription":"Internal server exception : Request to remote secondary @slow674 at null:null received error response 'AT0015-Exception: public:publickey@slow674 does not exist in keystore'"}
@jeremy_0 plookup:publickey@slow674
=> data:MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAl6HAhNfIC+gTPzVNk6ptm8rWLe2I2nDhktEFGlFUuNaHF7T4lS+n9zVLApKbwL/tXEMJAFZ2B7YhLJh4BILsAxsjMqTUuiMNVzpuWKEr3o06Nr7VBzu1n0PVcHdB1akj/pHhk60D7HWzqN2FZBYS0h1rYeVjpDVCh/WxPtqg/0f7sEv4uWpsWFivQVh7UPp5zB8fAj2ypmSyEvuMDU63N8pUUSmAWr7qKbuE+4Jj8QPY7C6CBBHjukMd5+oqkKGSqEkSn8enc22KLaFJiXkKKbqxVhLHBMCLsUl+1krxeFNU2PmTAcc6Ek+yby7XM0erhy99vbkaAC/cCuSeSS2qjQIDAQAB
```

**Comment by gkc** [comment]
> @cconstab For sshnpd, if the sshnp atSign does not exist, do we want to exit(1), or do a while() {sleep 5} or something?

@JeremyTubongbanua - @cconstab and I discussed this last week ... we think a `while (can't fetch public key) {try; sleep}` is best

**Comment by JeremyTubongbanua** [comment]
@gkc Done in https://github.com/atsign-foundation/sshnoports/pull/211/commits/f6fb13309c2af073ff9bd715596cbd5aa81b6f96

**Comment by JeremyTubongbanua** [comment]
Here's the results of my testing as of the most recent commit:


Testing that `sshnp` does not continue, if the other atSign is alive, but hasn't been onboarded yet:
```sh
atsign@6d0c777a3b17:~/.local/bin$ ./sshnp -f @jeremy_0 -t @noblesnowbaboon -s id_ed25519.pub -d docker -h @rv_am
Unhandled exception:
sshnpd atSign @noblesnowbaboon is not activated.
#0      SSHNP.init (package:sshnoports/sshnp.dart:181)
<asynchronous suspension>
#1      main (file:///build/repo/bin/sshnp.dart:21)
<asynchronous suspension>
```

Testing that `sshnpd` keeps on trying until the other atSign's `public:publickey` exists

```sh
atsign@8ea01a4b8c7a:~/.local/bin$ ./sshnpd -a @smoothalligator -m @noblesnowbaboon -d docker -s -u -v
...
WARNING|2023-06-27 04:40:25.940721| sshnpd |Waiting for @noblesnowbaboon to be activated... 

WARNING|2023-06-27 04:40:31.130089| sshnpd |Waiting for @noblesnowbaboon to be activated... 

WARNING|2023-06-27 04:40:36.445747| sshnpd |Waiting for @noblesnowbaboon to be activated...
```

Here are the `sshnpd` logs once I activate the `sshnp` atSign (in this case, @noblesnowbaboon) using at_activate
```sh
WARNING|2023-06-27 04:42:52.471506| sshnpd |Waiting for @noblesnowbaboon to be activated... 

INFO|2023-06-27 04:42:52.899482|AbstractAtKeyEncryption (@smoothalligator)|Encrypted shared symmetric key for @smoothalligator not found in local storage 

INFO|2023-06-27 04:42:52.899658|AbstractAtKeyEncryption (@smoothalligator)|Deleting @noblesnowbaboon:shared_key@smoothalligator from LocalSecondary 

INFO|2023-06-27 04:42:52.900637|AbstractAtKeyEncryption (@smoothalligator)|Fetching shared symmetric key for @smoothalligator from atServer 

INFO|2023-06-27 04:42:52.961281|AbstractAtKeyEncryption (@smoothalligator)|Creating new shared symmetric key as @smoothalligator for @noblesnowbaboon 

INFO|2023-06-27 04:42:52.963822|AbstractAtKeyEncryption (@smoothalligator)|Deleting @noblesnowbaboon:shared_key@smoothalligator from RemoteSecondary 

INFO|2023-06-27 04:42:53.072124|AbstractAtKeyEncryption (@smoothalligator)|Storing new shared symmetric key to atServer 

INFO|2023-06-27 04:42:53.144331|AbstractAtKeyEncryption (@smoothalligator)|Storing new shared symmetric key to local storage 

INFO|2023-06-27 04:42:53.145216|AbstractAtKeyEncryption (@smoothalligator)|'Their' copy of shared symmetric key for @noblesnowbaboon not found in local storage - will check atServer 

INFO|2023-06-27 04:42:53.271397|AbstractAtKeyEncryption (@smoothalligator)|Saving 'their' copy of shared symmetric key for @noblesnowbaboon to atServer 

INFO|2023-06-27 04:42:53.339776|AbstractAtKeyEncryption (@smoothalligator)|Saving 'their' copy of shared symmetric key for @noblesnowbaboon to local storage 

INFO|2023-06-27 04:42:54.041067| sshnpd |SUCCESS:id: 270e510e-412e-4666-83f8-39866465410c status: NotificationStatusEnum.delivered atsign 

INFO|2023-06-27 04:42:55.103470|SyncService (@smoothalligator)|72717253|Returning serverCommitId 32573 

INFO|2023-06-27 04:42:55.168053|SyncService (@smoothalligator)|72717253|Returning serverCommitId 32573 

SEVERE|2023-06-27 04:42:55.281406|EncryptionUtil|Error while decrypting value: Invalid argument(s): Invalid or corrupted pad block 

WARNING|2023-06-27 04:42:55.281474|SyncService (@smoothalligator)|Exception occurred when setting conflict info for @noblesnowbaboon:shared_key@smoothalligator | Exception: Invalid argument(s): Invalid or corrupted pad block 

INFO|2023-06-27 04:42:55.367503|SyncService (@smoothalligator)|72717253|Inside syncComplete. syncRequest.requestSource : SyncRequestSource.system; syncRequest.onDone : Closure: (SyncResult) => void from Function '_onDone@118025363':. 

INFO|2023-06-27 04:42:55.427854|SyncService (@smoothalligator)|72717253|Returning serverCommitId 32576
```

Then in the `sshnp` container, we test sshing into the `sshnpd` container (one final test that it all still works):
```sh
./sshnp -f @noblesnowbaboon -t @smoothalligator -s id_ed25519.pub -d docker -h @rv_am
ssh -p 40873 atsign@localhost -i /atsign/.ssh/id_ed25519
...
atsign@8ea01a4b8c7a:~$ ps -a
  PID TTY          TIME CMD
   24 pts/0    00:00:00 bash
   26 pts/0    00:00:06 dart:sshnpd
   73 pts/1    00:00:00 ps
...

```

**Comment by JeremyTubongbanua** [comment]
> Here's the results of my testing as of the most recent commit:
> 
> Testing that `sshnp` does not continue, if the other atSign is alive, but hasn't been onboarded yet:
> 
> ```shell
> atsign@6d0c777a3b17:~/.local/bin$ ./sshnp -f @jeremy_0 -t @noblesnowbaboon -s id_ed25519.pub -d docker -h @rv_am
> Unhandled exception:
> sshnpd atSign @noblesnowbaboon is not activated.
> #0      SSHNP.init (package:sshnoports/sshnp.dart:181)
> <asynchronous suspension>
> #1      main (file:///build/repo/bin/sshnp.dart:21)
> <asynchronous suspension>
> ```
> 
> Testing that `sshnpd` keeps on trying until the other atSign's `public:publickey` exists
> 
> ```shell
> atsign@8ea01a4b8c7a:~/.local/bin$ ./sshnpd -a @smoothalligator -m @noblesnowbaboon -d docker -s -u -v
> ...
> WARNING|2023-06-27 04:40:25.940721| sshnpd |Waiting for @noblesnowbaboon to be activated... 
> 
> WARNING|2023-06-27 04:40:31.130089| sshnpd |Waiting for @noblesnowbaboon to be activated... 
> 
> WARNING|2023-06-27 04:40:36.445747| sshnpd |Waiting for @noblesnowbaboon to be activated...
> ```
> 
> Here are the `sshnpd` logs once I activate the `sshnp` atSign (in this case, @noblesnowbaboon) using at_activate
> 
> ```shell
> WARNING|2023-06-27 04:42:52.471506| sshnpd |Waiting for @noblesnowbaboon to be activated... 
> 
> INFO|2023-06-27 04:42:52.899482|AbstractAtKeyEncryption (@smoothalligator)|Encrypted shared symmetric key for @smoothalligator not found in local storage 
> 
> INFO|2023-06-27 04:42:52.899658|AbstractAtKeyEncryption (@smoothalligator)|Deleting @noblesnowbaboon:shared_key@smoothalligator from LocalSecondary 
> 
> INFO|2023-06-27 04:42:52.900637|AbstractAtKeyEncryption (@smoothalligator)|Fetching shared symmetric key for @smoothalligator from atServer 
> 
> INFO|2023-06-27 04:42:52.961281|AbstractAtKeyEncryption (@smoothalligator)|Creating new shared symmetric key as @smoothalligator for @noblesnowbaboon 
> 
> INFO|2023-06-27 04:42:52.963822|AbstractAtKeyEncryption (@smoothalligator)|Deleting @noblesnowbaboon:shared_key@smoothalligator from RemoteSecondary 
> 
> INFO|2023-06-27 04:42:53.072124|AbstractAtKeyEncryption (@smoothalligator)|Storing new shared symmetric key to atServer 
> 
> INFO|2023-06-27 04:42:53.144331|AbstractAtKeyEncryption (@smoothalligator)|Storing new shared symmetric key to local storage 
> 
> INFO|2023-06-27 04:42:53.145216|AbstractAtKeyEncryption (@smoothalligator)|'Their' copy of shared symmetric key for @noblesnowbaboon not found in local storage - will check atServer 
> 
> INFO|2023-06-27 04:42:53.271397|AbstractAtKeyEncryption (@smoothalligator)|Saving 'their' copy of shared symmetric key for @noblesnowbaboon to atServer 
> 
> INFO|2023-06-27 04:42:53.339776|AbstractAtKeyEncryption (@smoothalligator)|Saving 'their' copy of shared symmetric key for @noblesnowbaboon to local storage 
> 
> INFO|2023-06-27 04:42:54.041067| sshnpd |SUCCESS:id: 270e510e-412e-4666-83f8-39866465410c status: NotificationStatusEnum.delivered atsign 
> 
> INFO|2023-06-27 04:42:55.103470|SyncService (@smoothalligator)|72717253|Returning serverCommitId 32573 
> 
> INFO|2023-06-27 04:42:55.168053|SyncService (@smoothalligator)|72717253|Returning serverCommitId 32573 
> 
> SEVERE|2023-06-27 04:42:55.281406|EncryptionUtil|Error while decrypting value: Invalid argument(s): Invalid or corrupted pad block 
> 
> WARNING|2023-06-27 04:42:55.281474|SyncService (@smoothalligator)|Exception occurred when setting conflict info for @noblesnowbaboon:shared_key@smoothalligator | Exception: Invalid argument(s): Invalid or corrupted pad block 
> 
> INFO|2023-06-27 04:42:55.367503|SyncService (@smoothalligator)|72717253|Inside syncComplete. syncRequest.requestSource : SyncRequestSource.system; syncRequest.onDone : Closure: (SyncResult) => void from Function '_onDone@118025363':. 
> 
> INFO|2023-06-27 04:42:55.427854|SyncService (@smoothalligator)|72717253|Returning serverCommitId 32576
> ```
> 
> Then in the `sshnp` container, we test sshing into the `sshnpd` container (one final test that it all still works):
> 
> ```shell
> ./sshnp -f @noblesnowbaboon -t @smoothalligator -s id_ed25519.pub -d docker -h @rv_am
> ssh -p 40873 atsign@localhost -i /atsign/.ssh/id_ed25519
> ...
> atsign@8ea01a4b8c7a:~$ ps -a
>   PID TTY          TIME CMD
>    24 pts/0    00:00:00 bash
>    26 pts/0    00:00:06 dart:sshnpd
>    73 pts/1    00:00:00 ps
> ...
> ```

cc @gkc

**Review by xavierchanth — APPROVED** [review]

**Review by xavierchanth — APPROVED** [review]

**Review by gkc — APPROVED** [review]
Nice work @JeremyTubongbanua

**Pull request merged**